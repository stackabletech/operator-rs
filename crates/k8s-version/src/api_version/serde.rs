use std::str::FromStr;

use serde::{Deserialize, Serialize, de::Visitor};

use crate::ApiVersion;

impl<'de> Deserialize<'de> for ApiVersion {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        struct ApiVersionVisitor;

        impl Visitor<'_> for ApiVersionVisitor {
            type Value = ApiVersion;

            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                write!(formatter, "a valid Kubernetes API version")
            }

            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
            where
                E: serde::de::Error,
            {
                ApiVersion::from_str(v).map_err(serde::de::Error::custom)
            }
        }

        deserializer.deserialize_str(ApiVersionVisitor)
    }
}

impl Serialize for ApiVersion {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&self.to_string())
    }
}

#[cfg(feature = "serde")]
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn deserialize() {
        let _: ApiVersion =
            serde_yaml::from_str("extensions.k8s.io/v1alpha1").expect("api version is valid");
    }

    #[test]
    fn serialize() {
        let api_version =
            ApiVersion::from_str("extensions.k8s.io/v1alpha1").expect("api version is valid");
        assert_eq!(
            "extensions.k8s.io/v1alpha1\n",
            serde_yaml::to_string(&api_version).expect("api version must serialize")
        );
    }
}
