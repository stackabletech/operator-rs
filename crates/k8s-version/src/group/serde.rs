use std::str::FromStr;

use serde::{Deserialize, Serialize, de::Visitor};

use crate::Group;

impl<'de> Deserialize<'de> for Group {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        struct GroupVisitor;

        impl<'de> Visitor<'de> for GroupVisitor {
            type Value = Group;

            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                write!(formatter, "a valid Kubernetes API group")
            }

            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
            where
                E: serde::de::Error,
            {
                Group::from_str(v).map_err(serde::de::Error::custom)
            }
        }

        deserializer.deserialize_str(GroupVisitor)
    }
}

impl Serialize for Group {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&self)
    }
}
