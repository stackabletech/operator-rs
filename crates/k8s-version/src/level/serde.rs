use std::str::FromStr;

use serde::{Deserialize, Serialize, de::Visitor};

use crate::Level;

impl<'de> Deserialize<'de> for Level {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        struct LevelVisitor;

        impl<'de> Visitor<'de> for LevelVisitor {
            type Value = Level;

            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                write!(formatter, "a valid Kubernetes API version level")
            }

            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
            where
                E: serde::de::Error,
            {
                Level::from_str(v).map_err(serde::de::Error::custom)
            }
        }

        deserializer.deserialize_str(LevelVisitor)
    }
}

impl Serialize for Level {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&self.to_string())
    }
}

#[cfg(feature = "serde")]
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn deserialize() {
        let _: Level = serde_yaml::from_str("alpha1").expect("level is valid");
    }

    #[test]
    fn serialize() {
        let api_version = Level::from_str("alpha1").expect("level is valid");
        assert_eq!(
            "alpha1\n",
            serde_yaml::to_string(&api_version).expect("level must serialize")
        );
    }
}
