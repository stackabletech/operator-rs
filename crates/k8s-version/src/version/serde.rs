use std::str::FromStr;

use serde::{Deserialize, Serialize, de::Visitor};

use crate::Version;

impl<'de> Deserialize<'de> for Version {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        struct VersionVisitor;

        impl<'de> Visitor<'de> for VersionVisitor {
            type Value = Version;

            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                write!(formatter, "a valid Kubernetes API version")
            }

            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
            where
                E: serde::de::Error,
            {
                Version::from_str(v).map_err(serde::de::Error::custom)
            }
        }

        deserializer.deserialize_str(VersionVisitor)
    }
}

impl Serialize for Version {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&self.to_string())
    }
}
