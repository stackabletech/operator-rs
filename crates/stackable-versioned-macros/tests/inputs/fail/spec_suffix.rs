use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use stackable_versioned_macros::versioned;

#[versioned(
    version(name = "v1alpha1"),
    version(name = "v1beta1"),
    version(name = "v1")
)]
#[derive(Clone, Debug, Deserialize, Serialize, JsonSchema)]
mod versioned {
    #[versioned(crd(group = "stackable.tech"))]
    pub struct Foo {
        #[versioned(
            added(since = "v1beta1"),
            changed(since = "v1", from_name = "bah", from_type = "u16")
        )]
        bar: usize,
        baz: bool,
    }
}

fn main() {}
