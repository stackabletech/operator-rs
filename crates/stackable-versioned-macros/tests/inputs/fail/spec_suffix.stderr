error: struct name needs to include the `Spec` suffix if CRD features are enabled via `#[versioned(crd())]`
  --> tests/inputs/fail/spec_suffix.rs:13:16
   |
13 |     pub struct Foo {
   |                ^^^
